File: windows/Scanner/RulePack/RulePackLoader.cs
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace AIExposureScanner.Scanner.RulePacks;

public abstract record RulePackLoadResult
{
    public sealed record Valid(RulePack Pack) : RulePackLoadResult;
    public sealed record Invalid(string Error) : RulePackLoadResult;
}

public static class RulePackLoader
{
    private static readonly IDeserializer Deserializer = new DeserializerBuilder()
        .WithNamingConvention(UnderscoredNamingConvention.Instance)
        .IgnoreUnmatchedProperties()
        .Build();

    public static RulePackLoadResult Load(string yaml)
    {
        try
        {
            return Parse(yaml);
        }
        catch (Exception ex)
        {
            return new RulePackLoadResult.Invalid($"YAML parse error: {ex.Message}");
        }
    }

    private static RulePackLoadResult Parse(string yaml)
    {
        var dto = Deserializer.Deserialize<RulePackDto?>(yaml);
        if (dto is null)
            return new RulePackLoadResult.Invalid("Empty YAML document");

        if (string.IsNullOrWhiteSpace(dto.Version))
            return new RulePackLoadResult.Invalid("Missing required field: version");
        if (string.IsNullOrWhiteSpace(dto.Id))
            return new RulePackLoadResult.Invalid("Missing required field: id");
        if (dto.Id.StartsWith("AES-", StringComparison.OrdinalIgnoreCase))
            return new RulePackLoadResult.Invalid("Rule pack id must not start with 'AES-' (reserved for built-in rules)");
        if (string.IsNullOrWhiteSpace(dto.Name))
            return new RulePackLoadResult.Invalid("Missing required field: name");

        var rules = new List<PackRule>();
        foreach (var (i, r) in (dto.Rules ?? []).Select((r, i) => (i, r)))
        {
            var err = ValidateRule(r, $"rules[{i}]");
            if (err is not null) return new RulePackLoadResult.Invalid(err);
            rules.Add(new PackRule(r.Id!, r.Severity!, r.Title ?? "", r.Explanation ?? "", r.Recommendation ?? "",
                new PackRuleMatch(r.Match!.Fact!,
                    r.Match.CommandEquals, r.Match.CommandContains,
                    r.Match.NameContains, r.Match.ArgsContain,
                    r.Match.App, r.Match.Key,
                    r.Match.ValueEquals, r.Match.ExtensionIdContains)));
        }

        var overrides = new List<PackOverride>();
        foreach (var (i, o) in (dto.Overrides ?? []).Select((o, i) => (i, o)))
        {
            var err = ValidateOverride(o, $"overrides[{i}]");
            if (err is not null) return new RulePackLoadResult.Invalid(err);
            overrides.Add(new PackOverride(o.Id!, o.Severity!));
        }

        var escalations = new List<PackEscalation>();
        foreach (var (i, e) in (dto.Escalations ?? []).Select((e, i) => (i, e)))
        {
            var err = ValidateEscalation(e, $"escalations[{i}]");
            if (err is not null) return new RulePackLoadResult.Invalid(err);
            escalations.Add(new PackEscalation(e.Requires!, e.Scope!, e.EscalateTo!, e.Reason ?? ""));
        }

        return new RulePackLoadResult.Valid(
            new RulePack(dto.Version!, dto.Id!, dto.Name!, rules, overrides, escalations));
    }

    // MARK: - Validation helpers

    private static readonly HashSet<string> ValidSeverities =
        new(["critical", "high", "medium", "low"], StringComparer.OrdinalIgnoreCase);

    private static readonly HashSet<string> ValidFacts =
        new(["mcp_server", "setting", "extension", "auth_file"], StringComparer.OrdinalIgnoreCase);

    private static readonly HashSet<string> ValidScopes =
        new(["per_server", "global"], StringComparer.OrdinalIgnoreCase);

    private static string? ValidateRule(RuleDto r, string prefix)
    {
        if (string.IsNullOrWhiteSpace(r.Id)) return $"{prefix}: missing required field 'id'";
        if (r.Id.StartsWith("AES-", StringComparison.OrdinalIgnoreCase))
            return $"{prefix}: rule id '{r.Id}' must not start with 'AES-'";
        if (!ValidSeverities.Contains(r.Severity ?? ""))
            return $"{prefix}: missing or invalid 'severity' (use: critical, high, medium, low)";
        if (r.Match is null) return $"{prefix}: missing required field 'match'";
        if (!ValidFacts.Contains(r.Match.Fact ?? ""))
            return $"{prefix}.match: missing or invalid 'fact' (use: mcp_server, setting, extension, auth_file)";
        return null;
    }

    private static string? ValidateOverride(OverrideDto o, string prefix)
    {
        if (string.IsNullOrWhiteSpace(o.Id)) return $"{prefix}: missing required field 'id'";
        if (!ValidSeverities.Contains(o.Severity ?? ""))
            return $"{prefix}: missing or invalid 'severity'";
        return null;
    }

    private static string? ValidateEscalation(EscalationDto e, string prefix)
    {
        if (e.Requires is null || e.Requires.Count == 0)
            return $"{prefix}: missing or empty 'requires' (list of rule IDs)";
        if (!ValidScopes.Contains(e.Scope ?? ""))
            return $"{prefix}: missing or invalid 'scope' (use: per_server, global)";
        if (!ValidSeverities.Contains(e.EscalateTo ?? ""))
            return $"{prefix}: missing or invalid 'escalate_to' (use: critical, high, medium, low)";
        return null;
    }

    // MARK: - DTO types (YAML deserialization targets)

    private sealed class RulePackDto
    {
        public string? Version { get; set; }
        public string? Id { get; set; }
        public string? Name { get; set; }
        public List<RuleDto>? Rules { get; set; }
        public List<OverrideDto>? Overrides { get; set; }
        public List<EscalationDto>? Escalations { get; set; }
    }

    private sealed class RuleDto
    {
        public string? Id { get; set; }
        public string? Severity { get; set; }
        public string? Title { get; set; }
        public string? Explanation { get; set; }
        public string? Recommendation { get; set; }
        public MatchDto? Match { get; set; }
    }

    private sealed class MatchDto
    {
        public string? Fact { get; set; }
        public string? CommandEquals { get; set; }
        public string? CommandContains { get; set; }
        public string? NameContains { get; set; }
        public string? ArgsContain { get; set; }
        public string? App { get; set; }
        public string? Key { get; set; }
        public string? ValueEquals { get; set; }
        public string? ExtensionIdContains { get; set; }
    }

    private sealed class OverrideDto
    {
        public string? Id { get; set; }
        public string? Severity { get; set; }
    }

    private sealed class EscalationDto
    {
        public List<string>? Requires { get; set; }
        public string? Scope { get; set; }
        public string? EscalateTo { get; set; }
        public string? Reason { get; set; }
    }
}

File: macos/Scanner/Sources/Scanner/RulePack/RulePackLoader.swift
import Yams

public enum RulePackLoadResult: Sendable {
    case valid(RulePack)
    case invalid(String)
}

public enum RulePackLoader {
    public static func load(yaml: String) -> RulePackLoadResult {
        do {
            return .valid(try parse(yaml: yaml))
        } catch let e as ParseError {
            return .invalid(e.message)
        } catch {
            return .invalid("YAML parse error: \(error.localizedDescription)")
        }
    }

    // MARK: - Private

    private struct ParseError: Error {
        let message: String
    }

    private static func parse(yaml: String) throws -> RulePack {
        guard let parsed = try Yams.load(yaml: yaml) else {
            throw ParseError(message: "Empty YAML document")
        }
        guard let root = parsed as? [String: Any] else {
            throw ParseError(message: "Root must be a YAML mapping")
        }

        guard let version = root["version"] as? String, !version.isEmpty else {
            throw ParseError(message: "Missing required field: version")
        }
        guard let id = root["id"] as? String, !id.isEmpty else {
            throw ParseError(message: "Missing required field: id")
        }
        if id.uppercased().hasPrefix("AES-") {
            throw ParseError(message: "Rule pack id must not start with 'AES-' (reserved for built-in rules)")
        }
        guard let name = root["name"] as? String, !name.isEmpty else {
            throw ParseError(message: "Missing required field: name")
        }

        let rules: [PackRule] = try (root["rules"] as? [[String: Any]] ?? [])
            .enumerated().map { i, r in try parseRule(r, prefix: "rules[\(i)]") }

        let overrides: [PackOverride] = try (root["overrides"] as? [[String: Any]] ?? [])
            .enumerated().map { i, r in try parseOverride(r, prefix: "overrides[\(i)]") }

        let escalations: [PackEscalation] = try (root["escalations"] as? [[String: Any]] ?? [])
            .enumerated().map { i, r in try parseEscalation(r, prefix: "escalations[\(i)]") }

        return RulePack(version: version, id: id, name: name,
                        rules: rules, overrides: overrides, escalations: escalations)
    }

    private static func parseRule(_ raw: [String: Any], prefix: String) throws -> PackRule {
        guard let id = raw["id"] as? String, !id.isEmpty else {
            throw ParseError(message: "\(prefix): missing required field 'id'")
        }
        if id.uppercased().hasPrefix("AES-") {
            throw ParseError(message: "\(prefix): rule id '\(id)' must not start with 'AES-'")
        }
        guard let severityRaw = raw["severity"] as? String,
              let severity = Severity(rawValue: severityRaw) else {
            throw ParseError(message: "\(prefix): missing or invalid 'severity' (use: critical, high, medium, low)")
        }
        guard let matchRaw = raw["match"] as? [String: Any] else {
            throw ParseError(message: "\(prefix): missing required field 'match'")
        }
        return PackRule(
            id: id,
            severity: severity,
            title: raw["title"] as? String ?? "",
            explanation: raw["explanation"] as? String ?? "",
            recommendation: raw["recommendation"] as? String ?? "",
            match: try parseMatch(matchRaw, prefix: "\(prefix).match")
        )
    }

    private static func parseMatch(_ raw: [String: Any], prefix: String) throws -> PackRuleMatch {
        guard let factRaw = raw["fact"] as? String,
              let fact = PackRuleMatch.FactType(rawValue: factRaw) else {
            throw ParseError(message: "\(prefix): missing or invalid 'fact' (use: mcp_server, setting, extension, auth_file)")
        }
        return PackRuleMatch(
            fact: fact,
            commandEquals: raw["command_equals"] as? String,
            commandContains: raw["command_contains"] as? String,
            nameContains: raw["name_contains"] as? String,
            argsContain: raw["args_contain"] as? String,
            app: raw["app"] as? String,
            key: raw["key"] as? String,
            valueEquals: raw["value_equals"] as? String,
            extensionIdContains: raw["extension_id_contains"] as? String
        )
    }

    private static func parseOverride(_ raw: [String: Any], prefix: String) throws -> PackOverride {
        guard let id = raw["id"] as? String, !id.isEmpty else {
            throw ParseError(message: "\(prefix): missing required field 'id'")
        }
        guard let severityRaw = raw["severity"] as? String,
              let severity = Severity(rawValue: severityRaw) else {
            throw ParseError(message: "\(prefix): missing or invalid 'severity'")
        }
        return PackOverride(id: id, severity: severity)
    }

    private static func parseEscalation(_ raw: [String: Any], prefix: String) throws -> PackEscalation {
        guard let requires = raw["requires"] as? [String], !requires.isEmpty else {
            throw ParseError(message: "\(prefix): missing or empty 'requires' (list of rule IDs)")
        }
        guard let scope = raw["scope"] as? String, EscalationScope(yamlValue: scope) != nil else {
            throw ParseError(message: "\(prefix): missing or invalid 'scope' (use: per_server, global)")
        }
        guard let escalateTo = raw["escalate_to"] as? String, Severity(rawValue: escalateTo) != nil else {
            throw ParseError(message: "\(prefix): missing or invalid 'escalate_to' (use: critical, high, medium, low)")
        }
        return PackEscalation(
            requires: requires,
            scope: scope,
            escalateTo: escalateTo,
            reason: raw["reason"] as? String ?? ""
        )
    }
}
